tests/ui/array-slice-vec/driftsort-off-by-one-issue-136103.rs

```diff
@@ -0,0 +1,53 @@
+//@ run-pass
+// Ensures that driftsort doesn't crash under specific slice
+// length and memory size.
+// Based on the example given in https://github.com/rust-lang/rust/issues/136103.
+use std::cmp::Ordering;
+use std::hint::black_box;
+
+#[derive(Eq)]
+struct S(bool, [u8; 999_999]);
+
+// Don't compare the large array to make
+// the test run faster.
+
+impl PartialOrd for S {
+    #[inline]
+    fn partial_cmp(&self, other: &S) -> Option<Ordering> {
+        self.0.partial_cmp(&other.0)
+    }
+}
+
+impl Ord for S {
+    #[inline]
+    fn cmp(&self, other: &S) -> Ordering {
+        self.0.cmp(&other.0)
+    }
+}
+
+impl PartialEq for S {
+    #[inline]
+    fn eq(&self, other: &S) -> bool {
+        self.0 == other.0
+    }
+}
+
+fn main() {
```

Comment:
This is test case can easily run into stack overflows, 1MB values are just asking for trouble, in my testing such values blow up `into_iter` and easily `slice::sort`, 2MB values can't even be used with `Vec::push` - particularly so in debug builds - in combination with being needlessly noisy - the `Ord` impls only distract - I suggest using the minimal reproducer @orlp shared.